File: apps/frontend/src/types/cashFlow.ts
/**
 * Dot-Store V2.2 现金流分析类型定义
 */

/**
 * 三色安全指数
 */
export interface SafetyIndex {
  safety_score: number;
  safety_level: 'safe' | 'warning' | 'danger';
  color_code: string;
  message: string;
  factors: {
    cash_balance: number;
    liability: number;
    net_available: number;
    days_of_operation: number;
  };
}

/**
 * 现金流分析
 */
export interface CashFlowAnalysis {
  id: number;
  user_id: number;
  analysis_date: string;
  analysis_type: string;
  period_start: string;
  period_end: string;
  total_income: number;
  total_expense: number;
  net_cash_flow: number;
  avg_daily_income: number;
  avg_daily_expense: number;
  income_structure: Record<string, number>;
  expense_structure: Record<string, number>;
  health_score: number;
  risk_level: 'low' | 'medium' | 'high' | 'critical';
  recommendations?: string[];
  created_at: string;
}

/**
 * 现金流预测
 */
export interface CashFlowForecast {
  id: number;
  user_id: number;
  forecast_date: string;
  target_date: string;
  predicted_income: number;
  predicted_expense: number;
  predicted_balance: number;
  confidence_level: number;
  risk_alert: boolean;
  alert_message?: string;
  created_at: string;
}

/**
 * 现金流预测列表响应
 */
export interface CashFlowForecastListResponse {
  items: CashFlowForecast[];
  total: number;
}

/**
 * 风险预警
 */
export interface RiskAlert {
  id: number;
  user_id: number;
  alert_date: string;
  alert_level: 'low' | 'medium' | 'high' | 'critical';
  alert_type: string;
  message: string;
  suggestions?: string[];
  is_read: boolean;
  is_resolved: boolean;
  resolved_at?: string;
  created_at: string;
}

/**
 * 风险预警列表响应
 */
export interface RiskAlertListResponse {
  items: RiskAlert[];
  total: number;
  unread_count: number;
}

/**
 * 收入结构响应
 */
export interface IncomeStructureResponse {
  total_income: number;
  income_by_category: Record<string, number>;
  income_by_mode: Record<string, number>;
  income_trend: Array<{
    date: string;
    amount: number;
  }>;
  stability_score: number;
  growth_rate?: number;
}

/**
 * 支出结构响应
 */
export interface ExpenseStructureResponse {
  total_expense: number;
  expense_by_category: Record<string, number>;
  expense_by_behavior: Record<string, number>;
  expense_by_function: Record<string, number>;
  expense_trend: Array<{
    date: string;
    amount: number;
  }>;
  anomaly_detected: boolean;
}

/**
 * 盈亏平衡分析
 */
export interface BreakEvenAnalysis {
  break_even_point: number;
  current_revenue: number;
  fixed_cost: number;
  variable_cost_ratio: number;
  contribution_margin_ratio: number;
  safety_margin: number;
  safety_margin_ratio: number;
  status: 'profit' | 'loss' | 'break_even';
}

/**
 * 月度健康报告
 */
export interface MonthlyReport {
  year: number;
  month: number;
  period_start: string;
  period_end: string;
  total_income: number;
  total_expense: number;
  net_cash_flow: number;
  income_structure: IncomeStructureResponse;
  expense_structure: ExpenseStructureResponse;
  break_even: BreakEvenAnalysis;
  safety_index: SafetyIndex;
  health_score: number;
  risk_level: string;
  recommendations: string[];
  action_items: string[];
  comparison?: {
    cash_balance_change: number;
    inventory_value_change: number;
    prepaid_balance_change: number;
  };
}

/**
 * 收入波动归因
 */
export interface IncomeVolatility {
  volatility_score: number;
  trend: 'up' | 'down' | 'stable';
  change_percentage: number;
  factors: Array<{
    category: string;
    amount: number;
    percentage: number;
  }>;
  main_cause: string;
  suggestions: string[];
}

/**
 * 仪表盘数据
 */
export interface DashboardData {
  safety_index: SafetyIndex;
  current_balance: number;
  yesterday_change: number;
  monthly_income: number;
  monthly_expense: number;
  monthly_profit: number;
  recent_trend: Array<{
    date: string;
    income: number;
    expense: number;
  }>;
  active_alerts: Array<{
    id: number;
    alert_level: string;
    alert_type: string;
    message: string;
    created_at: string;
  }>;
  forecast_summary: {
    has_risk: boolean;
    predicted_balance_end: number;
    avg_daily_income: number;
    avg_daily_expense: number;
  };
}

/**
 * 用户偏好配置
 */
export interface UserPreference {
  id: number;
  user_id: number;
  weekly_reminder_enabled: boolean;
  weekly_reminder_day: number;
  weekly_reminder_time: string;
  monthly_report_enabled: boolean;
  monthly_report_day: number;
  monthly_report_time: string;
  risk_alert_enabled: boolean;
  risk_alert_threshold: 'low' | 'medium' | 'high';
  notification_channels: string[];
}

File: apps/api-server/app/schemas/cash_flow.py
"""
Dot-Store V2.2 现金流分析Schema
"""
from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, ConfigDict


class SafetyIndexResponse(BaseModel):
    """三色安全指数响应"""
    model_config = ConfigDict(from_attributes=True)
    
    safety_score: float = Field(..., description="安全指数分数(0-100)")
    safety_level: str = Field(..., description="安全等级: safe/warning/danger")
    color_code: str = Field(..., description="颜色代码")
    message: str = Field(..., description="用户友好提示语")
    factors: Dict[str, Any] = Field(..., description="计算因素")


class CashFlowForecastResponse(BaseModel):
    """现金流预测响应"""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    user_id: int
    forecast_date: date
    target_date: date
    predicted_income: Decimal
    predicted_expense: Decimal
    predicted_balance: Decimal
    confidence_level: Decimal
    risk_alert: bool
    alert_message: Optional[str] = None
    created_at: datetime


class CashFlowForecastListResponse(BaseModel):
    """现金流预测列表响应"""
    items: List[CashFlowForecastResponse]
    total: int


class IncomeStructureResponse(BaseModel):
    """收入结构分析响应"""
    total_income: Decimal = Field(..., description="总收入")
    income_by_category: Dict[str, Decimal] = Field(..., description="按分类的收入")
    income_by_mode: Dict[str, Decimal] = Field(..., description="按模式的收入")
    income_trend: List[Dict[str, Any]] = Field(..., description="收入趋势")
    stability_score: float = Field(..., description="收入稳定性评分")
    growth_rate: Optional[float] = Field(None, description="收入增长率")


class ExpenseStructureResponse(BaseModel):
    """成本结构分析响应"""
    total_expense: Decimal = Field(..., description="总支出")
    expense_by_category: Dict[str, Decimal] = Field(..., description="按分类的支出")
    expense_by_behavior: Dict[str, Decimal] = Field(..., description="按行为分类的支出")
    expense_by_function: Dict[str, Decimal] = Field(..., description="按功能分类的支出")
    expense_trend: List[Dict[str, Any]] = Field(..., description="支出趋势")
    anomaly_detected: bool = Field(..., description="是否检测到异常")


class BreakEvenAnalysisResponse(BaseModel):
    """盈亏平衡分析响应"""
    break_even_point: Decimal = Field(..., description="盈亏平衡点(销售额)")
    current_revenue: Decimal = Field(..., description="当前收入")
    fixed_cost: Decimal = Field(..., description="固定成本")
    variable_cost_ratio: float = Field(..., description="变动成本比率")
    contribution_margin_ratio: float = Field(..., description="边际贡献率")
    safety_margin: Decimal = Field(..., description="安全边际")
    safety_margin_ratio: float = Field(..., description="安全边际率")
    status: str = Field(..., description="状态: profit/loss/break_even")


class CashFlowAnalysisResponse(BaseModel):
    """现金流分析响应"""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    user_id: int
    analysis_date: date
    analysis_type: str
    period_start: date
    period_end: date
    total_income: Decimal
    total_expense: Decimal
    net_cash_flow: Decimal
    avg_daily_income: Decimal
    avg_daily_expense: Decimal
    income_structure: Dict[str, Any]
    expense_structure: Dict[str, Any]
    health_score: Decimal
    risk_level: str
    recommendations: Optional[List[str]] = None
    created_at: datetime


class MonthlyReportResponse(BaseModel):
    """月度健康报告响应"""
    year: int
    month: int
    period_start: date
    period_end: date
    
    total_income: Decimal
    total_expense: Decimal
    net_cash_flow: Decimal
    
    income_structure: IncomeStructureResponse
    expense_structure: ExpenseStructureResponse
    break_even: BreakEvenAnalysisResponse
    safety_index: SafetyIndexResponse
    
    health_score: float
    risk_level: str
    
    recommendations: List[str]
    action_items: List[str]
    
    comparison: Optional[Dict[str, Any]] = None


class IncomeVolatilityResponse(BaseModel):
    """收入波动归因响应"""
    volatility_score: float = Field(..., description="波动性评分")
    trend: str = Field(..., description="趋势: up/down/stable")
    change_percentage: float = Field(..., description="变化百分比")
    
    factors: List[Dict[str, Any]] = Field(..., description="影响因素")
    main_cause: str = Field(..., description="主要原因")
    
    suggestions: List[str] = Field(..., description="优化建议")


class CashFlowAnalyzeRequest(BaseModel):
    """现金流分析请求"""
    period_start: date = Field(..., description="分析周期开始日期")
    period_end: date = Field(..., description="分析周期结束日期")
    analysis_type: str = Field(default="monthly", description="分析类型")


class DashboardResponse(BaseModel):
    """仪表盘响应"""
    safety_index: SafetyIndexResponse = Field(..., description="安全指数")
    current_balance: Decimal = Field(..., description="当前余额")
    yesterday_change: Decimal = Field(..., description="昨日变动")
    monthly_income: Decimal = Field(..., description="本月收入")
    monthly_expense: Decimal = Field(..., description="本月支出")
    monthly_profit: Decimal = Field(..., description="本月利润")
    recent_trend: List[Dict[str, Any]] = Field(..., description="近期趋势")
    active_alerts: List[Dict[str, Any]] = Field(..., description="活跃预警")
    forecast_summary: Dict[str, Any] = Field(..., description="预测摘要")
